src/java.base/share/classes/java/security/Provider.java

```diff
@@ -1227,25 +1275,29 @@ private void parseLegacyPut(String name, String value) {
      */
     public Service getService(String type, String algorithm) {
         checkInitialized();
-
         // avoid allocating a new ServiceKey object if possible
         ServiceKey key = previousKey;
         if (key.matches(type, algorithm) == false) {
             key = new ServiceKey(type, algorithm, false);
             previousKey = key;
         }
+
         if (!serviceMap.isEmpty()) {
             Service s = serviceMap.get(key);
             if (s != null) {
                 return s;
             }
         }
-        synchronized (this) {
-            ensureLegacyParsed();
-            if (legacyMap != null && !legacyMap.isEmpty()) {
-                return legacyMap.get(key);
+
+        if (!legacyMap.isEmpty()) {
```

Comment:
Looks like this check is redundant. get() will return `null` anyway.

Reply:
Hmm, I suppose so.